examples/nmf/script.jl

```diff
@@ -0,0 +1,347 @@
+# ```math
+# \min_{D \in \mathbb{R}^{m \times k}_{\ge 0}, \Lambda \in \mathbb{R}^{k \times n}_{\ge 0}} \sum_{i = 1}^{n} \operatorname{OT}_{\varepsilon}(X_i, (D\Lambda)_i) + \rho_1 E(D) + \rho_2 E(\Lambda),
+# ```
+# where $\operatorname{OT}_{\varepsilon}(\alpha, \beta)$ is the entropy-regularised optimal transport loss between two probability distributions $\alpha, \beta$ for a cost $C$, and $E$ is an entropy barrier function (a smooth approximation to the non-negativity constraint),
+# ```math
+# E(A) = \sum_{ij} (A_{ij} \log(A_{ij}) - 1).  
+# ```
+# The parameters $\rho_1, \rho_2$ control how "sharp" the non-negativity constraints are. As $\rho_1, \rho_2 \to 0$, the smoothed constraint approaches the hard non-negativity constraint. Finally, $\varepsilon$ controls the regularisation level for the optimal transport loss. 
+# 
+# This example shows how this method can be implemented using functions from the `Dual` submodule of OptimalTransport.jl. 
+#  
+# ## Set up prerequisites
+# 
+# Load packages that will be required later.
+#
+using OptimalTransport
+import OptimalTransport.Dual: Dual
+using MLDatasets: MLDatasets
+using StatsBase
+using Plots
+using LogExpFunctions
+using NNlib: NNlib
+using LinearAlgebra
+using Distances
+using Base.Iterators
+using NMF
+using Optim
+# 
+# Define `simplex_norm!`, which normalises `x` to sum to `1` along `dims`. 
+#
+function simplex_norm!(x; dims=1)
+    return x .= x ./ sum(x; dims=dims)
+end
+# 
+# ## Implementation
+#
+# We now implement Wasserstein-NMF. Rolet et al. split the original non-convex problem into a pair of convex problems, one for $D$ and one for $\Lambda$. 
+# ```math
+# \begin{aligned}
+# &\min_{D \in \mathbb{R}^{m \times k}} \sum_{i = 1}^{n} \operatorname{OT}_{\varepsilon}(X_i, (D\Lambda)_i) + \rho_2 E(D), \\ 
+# &\min_{\Lambda \in \mathbb{R}^{k \times n}} \sum_{i = 1}^{n} \operatorname{OT}_{\varepsilon}(X_i, (D\Lambda)_i) + \rho_1 E(\Lambda). 
+# \end{aligned}
+# ```
+# For each of these problems, the dual problem can be derived (see Section 3.3 of Rolet et al. for details on how this is done). These turn out to be 
+# ```math 
+# \begin{aligned}
+# &\min_{G \in \mathbb{R}^{m \times n}} \sum_{i = 1}^n \operatorname{OT}_{\varepsilon}^*(X_i, G_i) + \rho_2 \sum_{i = 1}^{n} E^*(-(G \Lambda^\top)_i / \rho_2) \\ 
+# &\min_{G \in \mathbb{R}^{m \times n}} \sum_{i = 1}^n \operatorname{OT}_{\varepsilon}^*(X_i, G_i) + \rho_1 \sum_{i = 1}^{n} E^*(-(D^\top G)_i / \rho_1). 
+# \end{aligned}
+# ```
+# The semi-dual of entropy-regularised optimal transport loss, $\operatorname{OT}_{\varepsilon}$, is implemented as `Dual.ot_entropic_semidual` and its gradient can be computed by `Dual.ot_entropic_semidual_grad`. $E^*$ turns out to be `logsumexp`, for which we implement a wrapper function:
+#
+function E_star(x; dims=1)
```

Comment:
Maybe it's a bit more explicit without the wrapper? Or alternatively maybe use unicode characters?

Reply:
The motivation for using the wrapper function instead of just logsumexp is for better readability. With unicode, unfortunately `*` and `⋆` don't seem to be allowed in variable names...